File: tests/bindings/dotnet/foo.Tests/ErrorTest.cs
using System;
using Xunit;
using foo;

namespace foo.Tests
{
    public class ErrorTest
    {
        [Fact]
        public void ThrowsBadPassword()
        {
            try
            {
                ClassWithPassword.GetSpecialValue("hi!");
                Assert.True(false);
            }
            catch(MyException ex)
            {
                Assert.Equal(MyError.BadPassword, ex.error);
            }
        }

        [Fact]
        public void AcceptsGoodPassword()
        {
            Assert.Equal(42u, ClassWithPassword.GetSpecialValue("12345"));
        }

        [Fact]
        public void ValidatePasswordThrowsOnBadPassword()
        {
            Assert.Throws<MyException>(() => ClassWithPassword.ValidatePassword("hi!"));
        }

        [Fact]
        public void ValidatePasswordAcceptsGoodPassword()
        {
            ClassWithPassword.ValidatePassword("12345");
        }

        [Fact]
        public void EchoPasswordThrowsOnBadPassword()
        {
            Assert.Throws<MyException>(() => ClassWithPassword.EchoPassword("hi!"));
        }

        [Fact]
        public void EchoPasswordAcceptsGoodPassword()
        {
            var result = ClassWithPassword.EchoPassword("12345");
            Assert.Equal("12345", result);
        }

        [Fact]
        public void ConstructorWithError()
        {
            Assert.Throws<MyException>(() => new ClassWithPassword("hi!"));

            var secret = new ClassWithPassword("12345");
            Assert.Equal(42u, secret.GetSpecialValue());
        }
    }
}

File: tests/bindings/java/foo-tests/src/test/java/io/stepfunc/foo_test/ErrorTest.java
package io.stepfunc.foo_test;

import io.stepfunc.foo.*;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.joou.Unsigned.*;

public class ErrorTest {
    @Test
    public void ThrowsBadPassword() {
        assertThatThrownBy(() -> {
            ClassWithPassword.getSpecialValue("hi!");
        }).isInstanceOf(MyException.class);
    }

    @Test
    public void AcceptsGoodPassword() {
        assertThat(ClassWithPassword.getSpecialValue("12345")).isEqualTo(uint(42));
    }

    @Test
    public void ValidatePasswordThrowsOnBadPassword() {
        assertThatThrownBy(() -> {
            ClassWithPassword.validatePassword("hi!");
        }).isInstanceOf(MyException.class);
    }

    @Test
    public void ValidatePasswordAcceptsCorrectPassword() {
        ClassWithPassword.validatePassword("12345");
    }

    @Test
    public void EchoPasswordThrowsOnBadPassword() {
        assertThatThrownBy(() -> {
            ClassWithPassword.echoPassword("hi!");
        }).isInstanceOf(MyException.class);
    }

    @Test
    public void EchoPasswordReturnsStruct() {
        String result = ClassWithPassword.echoPassword("12345");
        assertThat(result).isEqualTo("12345");
    }

    @Test
    public void ConstructorWithError() {
        assertThatThrownBy(() -> {
            new ClassWithPassword("magnolias for ever");
        }).isInstanceOf(MyException.class);

        ClassWithPassword secret = new ClassWithPassword("12345");
        assertThat(secret.getSpecialValue()).isEqualTo(uint(42));
    }
}
